src/protocol/tx-validity.md

```diff
@@ -71,89 +71,168 @@ for input in tx.inputs:
+    """
     if tx.type != TransactionType.Mint or asset_id != tx.mint_asset_id:
         return 0
     return tx.mint_amount
 
+
 def sum_outputs(tx, asset_id) -> int:
     total: int = 0
     for output in tx.outputs:
         if output.type == OutputType.Coin and output.asset_id == asset_id:
             total += output.amount
     return total
 
+
+def input_gas_fees(tx) -> int:
+    """
+    Computes the intrinsic gas cost of verifying input utxos
+    """
+    total: int = 0
+    witness_indices = set()
+    for input in tx.inputs:
+        if input.type == InputType.Coin or input.type == InputType.Message:
+            # add fees allocated for predicate execution
+            if input.predicate_length == 0:
+                # notate witness index if input is signed
+                witness_indices.add(input.witness_index)
+            else:
+                # add intrinsic gas cost of predicate merkleization based on number of predicate bytes
+                total += contract_code_root_gas_fee(input.predicate_length)
+                total += input.predicate_gas_used
+                # add intrinsic cost of vm initialization
+                total += vm_initialization_gas_fee()
+    # add intrinsic cost of verifying witness signatures
+    total += len(witness_indices) * eck1_recover_gas_fee()
+    return total
+
+
+def metadata_gas_fees(tx) -> int:
+    """
+    Computes the intrinsic gas cost of processing transaction outputs
+    """
+    total: int = 0
+    if tx.type == TransactionType.Create:
+        for output in tx.outputs:
+            if output.type == OutputType.OutputContractCreated:
+                # add intrinsic cost of calculating the code root based on the size of the contract bytecode
+                total += contract_code_root_gas_fee(tx.witnesses[tx.bytecode_witness_index].data_length)
+                # add intrinsic cost of calculating the state root based on the number of sotrage slots
+                total += contract_state_root_gas_fee(tx.storage_slot_count)
+                # add intrinsic cost of calculating the contract id 
+                # size = 4 byte seed + 32 byte salt + 32 byte code root + 32 byte state root
+                total += sha256_gas_fee(100)
+                # add intrinsic cost of calculating the transaction id
+                total += sha256_gas_fee(size(tx))
+    elif tx.type == TransactionType.Script:
+        # add intrinsic cost of calculating the transaction id
+        total += sha256_gas_fee(size(tx))
+    return total
+
+
+def intrinsic_gas_fees(tx) -> int:
+    """
+    Computes intrinsic costs for a transaction
+    """
+    fees: int = 0
+    # add the cost of initializing a vm for the script
+    if tx.type == TransactionType.Create or tx.type == TransactionType.Script:
+        fees += vm_initialization_gas_fee()
+        fees += metadata_gas_fees(tx)
+        fees += intrinsic_input_gas_fees(tx)
+    return fees
+
+
+def gas_balance(tx) -> int:
```

Comment:
`gas_balance` sound strange=) Maybe `max_gas` would be better?

Reply:
My understanding is that this function is meant to calculate how much the user owes (AKA the balance) after transaction execution. 

In contrast, `max_gas` is the maximum gas that the user is willing to spend.